src/support/err.c

```diff
@@ -151,220 +203,211 @@ __wt_event_handler_set(WT_SESSION_IMPL *session, WT_EVENT_HANDLER *handler)
-        size_t __len;                                                   \
-        WT_ERR(__wt_vsnprintf_len_set(p, remain, &__len, __VA_ARGS__)); \
-        if (__len > remain)                                             \
-            __len = remain;                                             \
-        p += __len;                                                     \
-        remain -= __len;                                                \
-    } while (0)
-
-/*
- * __eventv_unpack_json_str --
- *     Unpack a string into JSON escaped format. Can be called with null destination for sizing.
- */
-static size_t
-__eventv_unpack_json_str(u_char *dest, size_t dest_len, char *src, size_t src_len)
-  WT_GCC_FUNC_ATTRIBUTE((cold))
-{
-    size_t len, n, nchars;
-    u_char *q;
-    const char *p;
-
-    nchars = 0;
-    q = dest;
-
-    for (p = src, len = src_len; len > 0; p++, --len) {
-        n = __wt_json_unpack_char((u_char)*p, q, dest_len, false);
-        nchars += n;
-        if (q != NULL) {
-            dest_len -= n;
-            q += n;
-        }
-    }
-
-    if (q != NULL)
-        *q = '\0';
-
-    return (nchars);
-}
 
 /*
- * __eventv_gen_msg --
- *     Generate a formatted message.
+ * __eventv --
+ *     Report a message to an event handler.
  */
 static int
-__eventv_gen_msg(WT_SESSION_IMPL *session, char *buffer, size_t *remain, bool is_json, int error,
-  const char *func, int line, WT_VERBOSE_CATEGORY category, WT_VERBOSE_LEVEL level, const char *msg)
+__eventv(WT_SESSION_IMPL *session, bool is_json, int error, const char *func, int line,
+  WT_VERBOSE_CATEGORY category, WT_VERBOSE_LEVEL level, const char *fmt, va_list ap)
   WT_GCC_FUNC_ATTRIBUTE((cold))
 {
     struct timespec ts;
+    WT_DECL_ITEM(json_msg);
+    WT_DECL_ITEM(tmp);
     WT_DECL_RET;
-    size_t len, msg_len, remain_msg, unpacked_msg_len;
-    u_char *unpacked_json_str;
-    char msg_str[2 * 1024], *p, *p_msg, tid[128];
+    WT_EVENT_HANDLER *handler;
+    WT_SESSION *wt_session;
+    size_t len, prefix_len, remain;
+    char *final, *p, tid[128];
     const char *err, *prefix, *verbosity_level_tag;
+    bool no_stderr;
+    va_list ap_copy;
 
-    p = buffer;
-    p_msg = msg_str;
-    unpacked_json_str = NULL;
+    /* SECURITY: Message buffer placed at the end of the stack in case snprintf overflows. */
+    char msg[4 * 1024];
 
-    remain_msg = sizeof(msg_str);
+    /*
+     * This function MUST handle a NULL WT_SESSION_IMPL handle.
+     *
+     * Without a session, we don't have event handlers or prefixes for the error message. Write the
+     * error to stderr and call it a day. (It's almost impossible for that to happen given how early
+     * we allocate the first session, but if the allocation of the first session fails, for example,
+     * we can end up here without a session.)
+     */
+    if (session == NULL)
+        return (__eventv_stderr(error, func, line, fmt, ap));
```

Comment:
If we end up here, is it intended to ignore the `is_json` field?

Reply:
Yes; `__eventv_stderr()` just dumps what we have to stderr.

The paths where I expect to see it actually happen is if we don't have a `WT_SESSION_IMPL` handle or we fail to allocate the memory required to format a huge message. There are other paths (imagine `snprintf` failed for some reason?), but those are the two where I think it might actually fire.

We could do better, but it would be non-trivial: for example, there's no `WT_SESSION.error_prefix` field and we'd have to consider that, and we can't get a timestamp, so the JSON wouldn't have `ts_sec/ts_usec` fields.

I think just throwing up our hands and pushing flat text to `stderr` is what used to happen.

Let me know if you think this should change.

Reply:
I am ok with this approach. I just realized we are about to do the same in the MongoDB layer, if outputting JSON fails, we fall back on logging a simple flat string.